### ISSUE TITLE
Smallrye GraphQL's parallel execution not compatible with Hibernate reactive

### ISSUE THREAD

**Issue by eamelink** [quarkusio/quarkus#32870]
### Describe the bug

When creating a GraphQL API with data retrieved with Hibernate reactive, I get the following exception:

`2023-04-24 21:14:48,145 ERROR [io.sma.graphql] (vert.x-eventloop-thread-1) SRGQL012000: Data Fetching Error: java.lang.IllegalStateException: Illegal pop() with non-matching JdbcValuesSourceProcessingState`

Note that this is very similar to the problem in #32790, but I think here it's caused by the underlying library `graphql-java` doing data fetching in parallel (@phillip-kruger correct me if I'm wrong).

### Expected behavior

I would expect a straightforward GraphQL query to work without having to think too much about Hibernate Reactive Sessions

### Actual behavior

An exception is thrown

### How to Reproduce?

Here's a reproducer: https://github.com/eamelink/graphql-and-hibernate-reactive

### Output of `uname -a` or `ver`

Darwin Eriks-MacBook-Pro.local 22.4.0 Darwin Kernel Version 22.4.0: Mon Mar  6 20:59:28 PST 2023; root:xnu-8796.101.5~3/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

openjdk version "17.0.6" 2023-01-17 OpenJDK Runtime Environment Temurin-17.0.6+10 (build 17.0.6+10) OpenJDK 64-Bit Server VM Temurin-17.0.6+10 (build 17.0.6+10, mixed mode)

### Quarkus version or git rev

3.0.0.CR2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.8 (4c87b05d9aedce574290d1acc98575ed5eb6cd39) Maven home: /Users/eamelink/.m2/wrapper/dists/apache-maven-3.8.8-bin/67c30f74/apache-maven-3.8.8 Java version: 17.0.6, vendor: Eclipse Adoptium, runtime: /Users/eamelink/.sdkman/candidates/java/17.0.6-tem Default locale: en_NL, platform encoding: UTF-8 OS name: "mac os x", version: "13.3.1", arch: "aarch64", family: "mac"

**Comment by eamelink** [comment]
Strangely enough, it doesn't appear to be a problem of shared sessions; the GraphQL engine does independent calls to the `Service` class where the `@WithSession` annotation is. So they should all have their own session and not conflict.

But if I put my debugger in `io.quarkus.hibernate.reactive.panache.common.runtime.SessionOperations`, on [these lines](https://github.com/quarkusio/quarkus/blob/1bc23e43d5b1435d7c1338c0ff0581b7bdea598c/extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java#L120-L122):

```java
return getSessionFactory()
        .openSession()
        .invoke(s -> context.putLocal(key, s))
```

I see that I get two invocations of `getSessionFactory()` from the same vert-x thread with the same vert-x context **before** two invocations of `s -> context.putLocal(key, s)`

So maybe just a race condition in this code?

**Comment by Sanne** [comment]
@eamelink many thanks for the report, that's indeed very suspicious.

**Comment by Sanne** [comment]
@phillip-kruger I've sent a PR to improve the code in Panache Reactive to help spotting code which is creating an illegal chain of operations: #32917 

But I suspect, looking into @eamelink 's reproducer, that the mistake is happening somewhere in graphql's integration code. Could you have a look please? 

I'm a bit surprised this is happening on a duplicated context, it would seem this is being marked as "safe" when it's not.

**Comment by DavideD** [comment]
I think it's happening here: https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/execution/AsyncExecutionStrategy.java#L48

GraphQL expect to be able to run queries in parallele and, because of this, some operations  happen in parallel trying to create a new session (in the example, when loading the foos).

Using the HIbenrate Reactive factory directly, seems to work as a workaround:
```

@ApplicationScoped
public class PeopleService {

    @Inject
    Mutiny.SessionFactory sf;

    public Uni<List<Person>> people() {
        return sf.withTransaction( session -> session
                .createQuery("from Person", Person.class )
                .getResultList() );
    }

    public Uni<List<Foo>> foos(Person person) {
        return sf.withTransaction( session -> session
                .createQuery( "from Foo where personId = :personId", Foo.class )
                .setParameter( "personId", person.id )
                .getResultList() );
    }
}
```


On a side note, you also need to change the import.sql to:
```
INSERT into Person (id, name) VALUES (1, 'Erik');
INSERT into Person (id, name) VALUES (2, 'Willem Jan');

INSERT into Foo (id, personId, foo) VALUES(1, 1, 'Eriks Foo');
INSERT into Foo (id, personId, foo) VALUES(2, 1, 'Eriks 2nd Foo');
INSERT into Foo (id, personId, foo) VALUES(3, 1, 'Eriks 3rd Foo');

INSERT into Foo (id, personId, foo) VALUES(4, 2, 'Willems Foo');
INSERT into Foo (id, personId, foo) VALUES(5, 2, 'Willems 2nd Foo');
```

(You need to specify the id for each Foo)

**Comment by Sanne** [comment]
@DavideD nice find, right that code seems highly supsicious. @phillip-kruger we'll need to patch that, perhaps it's easier to override with a custom implementation in Quarkus?  I'll leave it to you... :)

**Comment by mkouba** [comment]
> Using the Hibernate Reactive factory directly, seems to work as a workaround:

@DavideD Hm, but `SessionFactory.withTransaction()` does reuse the existing reactive session, or? I don't see a difference compared to `@WithTransaction`...

**Comment by DavideD** [comment]
>  @DavideD Hm, but SessionFactory.withTransaction() does reuse the existing reactive session, or? I don't see a difference compared to @WithTransaction...

Yes, I'm not sure why it works. It might be that it's just a bit faster and doesn't cause the issue.

**Comment by phillip-kruger** [comment]
Doing requests (read) in parallel is part of the GraphQL Spec (https://spec.graphql.org/June2018/#sec-Query). 
@jmartisk - do you have capacity to look at this ?

**Comment by jmartisk** [comment]
Sorry I don't really know how to fix it, this is far beyond my knowledge of the reactive world.

**Comment by paomian** [comment]
if i transfrom Uni<List<T>> to Multi<T> and use `transformToUniAndMerge` will case same error. 
use `transformToUniAndConcatenate` will avoid this error.

**Comment by paomian** [comment]
if i transfrom Uni<List<T>> to Multi<T> and use `transformToUniAndMerge` will case same error. 
use `transformToUniAndConcatenate` will avoid this error.
it shard context in many worker thread I think like https://github.com/quarkusio/quarkus/issues/32790.

**Comment by phillip-kruger** [comment]
@paomian so you want to do a PR?

**Comment by paomian** [comment]
@phillip-kruger I don't think I'm good enough to do a PR.
I just report a bug.

**Comment by uPagge** [comment]
I faced the same problem, but when receiving REST requests. Translated the service from 2.16.6 to 3.0.1

**Comment by eamelink** [comment]
I wrote a bit on how we work around it: https://blog.lunatech.com/posts/2023-05-19-hibernate-reactive-gotchas

**Comment by Sanne** [comment]
@eamelink great writeup, thanks! You should join our brainstorming sessions on Hibernate's zulip ;)

We also discussed recently that this limitation needs to be improved on; similarly to what you write, I'm inclined to introduce a queue so to transparently process "parallel" operations in sequence.

The option to open multiple sessions is a no-go IMO; separate transactions - like you mention - are a significant problem, but it gets worse with undefined boundaries for the transactions and we also need to keep in mind resource restrictions, for example what happens when you run out of connections and can you still guarantee liveness properties and healthy progress of the business operations.

The short version is that I agree that just documenting such a limitation doesn't cut it; it was perhaps good enough in the early days as we were going fast with the prototype and we had bigger concerns, such as figuring out if the whole thing was viable, but now that mostly everything else is working as expected we need to return to this.

Thanks for the feedback!

**Comment by ThomasKluiters** [comment]
@eamelink Thank you so much for that blog-post, this issue was driving me insane.

Though, in your solution there was a slight issue.

Instead of:

```java
   @Override
    public <T> Uni<T> protect(Uni<T> uni) {
        return acquire().replaceWith(uni).eventually(this::release);
    }
```

You meant:

```java
    @Override
    public <T> Uni<T> protect(Supplier<Uni<T>> inner) {
        return acquire().flatMap(ignored -> inner.get()).eventually(this::release);
    }
```

**Comment by elrob** [comment]
This issue got me too.
Thanks for the blog post!

It feels like there needs to be something more in the type system to make this more manageable.
Sometimes the `Uni`s that need to be protected are futher down the stack and at the point in the code where e.g. a `Uni.combine().all(` is invoked, it can be difficult to be aware that it requires to `protecting`.

**Comment by uPagge** [comment]
Hey, everybody.
I am facing the same problem when using two @Source queries at the same time. So, if I call only one @Source query in a GraphQL query, everything works correctly, but if I call two @Source queries at once, I get an error.

**Comment by gsmet** [comment]
@jmartisk is it still an issue? I know you changed a few things in SmallRye GraphQL since then?

**Comment by jmartisk** [comment]
I don't recall any changes specifically addressing this

**Comment by nelsoncc** [comment]
I'm trying to executing something similar to this input

````graphql
query getFilms {
   film0: film(filmId: 0) { title }
   film1: film(filmId: 1) { title }
}
````

````java
@GraphQLApi
@RequiredArgsConstructor(onConstructor = @__(@Inject))
public class MoviesGraphQLResource {

private final Service service;

    @Query("getFilms")
    public Uni<PaginatedFilm> getFilm(String filmId,FilmFilter filter) {
        return service.getFilms(participantId, filter);
    }
}

@ApplicationScoped
@RequiredArgsConstructor(onConstructor = @__(@Inject))
public class MovieService {

private final MovieRepository repository;

@WithSession
public Uni<PaginatedMovies> getFilms(
return repository.findFiltered(id, filter)
        .flatMap(
        entities -> {
            Uni<Long> totalCountUni;
            if (entities.isEmpty()) {
                totalCountUni = Uni.createFrom().item(0L);
            } else {
                totalCountUni =
                        repository.countFiltered(id, filter);
            }
            // Chain the count Uni and combine results sequentially
            return totalCountUni.map(
                    totalCount -> {
                        return entities.isEmpty()
                                ? createEmptyPaginatedResponse(totalCount)
                                : mapToPaginatedResponse(
                                        entities, totalCount, page, size);
                    });
        });
}
}

In repository I use extend from PanacheRepositoryBase and use `find` and `count`
````

I've got the errors

````shell
2025-08-06T09:27:56.244 ERROR traceId=c041c682eddeaa0863c92dfba25e4c37, parentId=6fa635c1c444250e, spanId=05a689714bbd647c, sampled=true, userId=, sessionId=, [io.sm.graphql] (vert.x-eventloop-thread-4) SRGQL012000: Data Fetching Error: java.lang.IllegalStateException: Illegal pop() with non-matching JdbcValuesSourceProcessingState
        at org.hibernate.sql.results.spi.LoadContexts.deregister(LoadContexts.java:40)
        at org.hibernate.reactive.sql.results.spi.ReactiveListResultsConsumer.end(ReactiveListResultsConsumer.java:186)
        at org.hibernate.reactive.sql.results.spi.ReactiveListResultsConsumer.lambda$consume$4(ReactiveListResultsConsumer.java:110)
        at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.unroll(AsyncTrampoline.java:131)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.lambda$unroll$0(AsyncTrampoline.java:126)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
        at io.vertx.core.Future.lambda$toCompletionStage$3(Future.java:600)
        at io.vertx.core.impl.future.FutureImpl$4.onSuccess(FutureImpl.java:176)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
        at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:88)
        at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:32)
        at io.vertx.core.Promise.complete(Promise.java:66)
        at io.vertx.core.Promise.handle(Promise.java:51)
        at io.vertx.core.Promise.handle(Promise.java:29)
        at io.vertx.core.impl.future.FutureImpl$4.onSuccess(FutureImpl.java:176)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
        at io.vertx.core.impl.future.Transformation$1.onSuccess(Transformation.java:61)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
        at io.vertx.core.impl.future.SucceededFuture.addListener(SucceededFuture.java:88)
        at io.vertx.core.impl.future.Transformation.onSuccess(Transformation.java:42)
        at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:60)
        at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:312)
        at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:168)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:57)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
        at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
        at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41)
        at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:23)
        at io.vertx.sqlclient.impl.command.CommandResponse.fire(CommandResponse.java:46)
        at io.vertx.sqlclient.impl.SocketConnectionBase.handleMessage(SocketConnectionBase.java:324)
        at io.vertx.pgclient.impl.PgSocketConnection.handleMessage(PgSocketConnection.java:114)
        at io.vertx.sqlclient.impl.SocketConnectionBase.lambda$init$0(SocketConnectionBase.java:137)
        at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:342)
        at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:335)
        at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:389)
        at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:159)
        at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
        at io.vertx.pgclient.impl.codec.PgDecoder.fireCommandResponse(PgDecoder.java:52)
        at io.vertx.pgclient.impl.codec.PgCommandCodec.handleReadyForQuery(PgCommandCodec.java:133)
        at io.vertx.pgclient.impl.codec.PgDecoder.decodeReadyForQuery(PgDecoder.java:248)
        at io.vertx.pgclient.impl.codec.PgDecoder.channelRead(PgDecoder.java:107)
        at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:796)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:732)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:658)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1583)

2025-08-06T09:27:56.247 ERROR traceId=c041c682eddeaa0863c92dfba25e4c37, parentId=6fa635c1c444250e, spanId=05a689714bbd647c, sampled=true, userId=, sessionId=, [io.sm.graphql] (vert.x-eventloop-thread-4) SRGQL012000: Data Fetching Error: java.lang.IllegalStateException: Session/EntityManager is closed
        at org.hibernate.internal.AbstractSharedSessionContract.checkOpen(AbstractSharedSessionContract.java:515)
        at org.hibernate.engine.spi.SharedSessionContractImplementor.checkOpen(SharedSessionContractImplementor.java:185)
        at org.hibernate.reactive.session.impl.ReactiveSessionImpl.checkOpen(ReactiveSessionImpl.java:1751)
        at org.hibernate.internal.SessionImpl.checkOpenOrWaitingForAutoClose(SessionImpl.java:610)
        at org.hibernate.internal.SessionImpl.instantiate(SessionImpl.java:1511)
        at org.hibernate.sql.results.graph.entity.internal.EntityInitializerImpl.instantiateEntity(EntityInitializerImpl.java:1268)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.reactiveResolveEntityInstance(ReactiveEntityInitializerImpl.java:566)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.reactiveResolveEntityInstance2(ReactiveEntityInitializerImpl.java:525)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.reactiveResolveEntityInstance1(ReactiveEntityInitializerImpl.java:495)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.postAssembleId(ReactiveEntityInitializerImpl.java:287)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.reactiveResolveInstance(ReactiveEntityInitializerImpl.java:268)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.lambda$reactiveResolveKey$11(ReactiveEntityInitializerImpl.java:645)
        at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1187)
        at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2341)
        at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:144)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.reactiveResolveKey(ReactiveEntityInitializerImpl.java:640)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.reactiveResolveKey(ReactiveEntityInitializerImpl.java:619)
        at org.hibernate.reactive.sql.results.graph.entity.internal.ReactiveEntityInitializerImpl.reactiveResolveKey(ReactiveEntityInitializerImpl.java:57)
        at org.hibernate.reactive.sql.results.internal.ReactiveStandardRowReader.resolveKey(ReactiveStandardRowReader.java:383)
        at org.hibernate.reactive.sql.results.internal.ReactiveStandardRowReader.lambda$coordinateInitializers$28(ReactiveStandardRowReader.java:376)
        at org.hibernate.reactive.util.impl.CompletionStages.lambda$loop$8(CompletionStages.java:483)
        at org.hibernate.reactive.util.impl.CompletionStages$ArrayLoop.next(CompletionStages.java:557)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.lambda$asyncWhile$1(AsyncTrampoline.java:215)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.unroll(AsyncTrampoline.java:121)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline$TrampolineInternal.trampoline(AsyncTrampoline.java:102)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:197)
        at org.hibernate.reactive.util.async.impl.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:215)
        at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:484)
        at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:455)
        at org.hibernate.reactive.sql.results.internal.ReactiveStandardRowReader.coordinateInitializers(ReactiveStandardRowReader.java:376)
        at org.hibernate.reactive.sql.results.internal.ReactiveStandardRowReader.reactiveReadRow(ReactiveStandardRowReader.java:126)
        at org.hibernate.reactive.sql.results.spi.ReactiveListResultsConsumer.lambda$addToResultsSupplier$6(ReactiveListResultsConsumer.java:160)
        at org.hibernate.reactive.sql.results.spi.ReactiveListResultsConsumer.lambda$consume$1(ReactiveListResultsConsumer.java:97)
        at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
        at io.vertx.core.Future.lambda$toCompletionStage$3(Future.java:600)
        at io.vertx.core.impl.future.FutureImpl$4.onSuccess(FutureImpl.java:176)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
        at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:88)
        at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:32)
        at io.vertx.core.Promise.complete(Promise.java:66)
        at io.vertx.core.Promise.handle(Promise.java:51)
        at io.vertx.core.Promise.handle(Promise.java:29)
        at io.vertx.core.impl.future.FutureImpl$4.onSuccess(FutureImpl.java:176)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
        at io.vertx.core.impl.future.Transformation$1.onSuccess(Transformation.java:61)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:66)
        at io.vertx.core.impl.future.SucceededFuture.addListener(SucceededFuture.java:88)
        at io.vertx.core.impl.future.Transformation.onSuccess(Transformation.java:42)
        at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:60)
        at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:312)
        at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:168)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:57)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:259)
        at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
        at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41)
        at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:23)
        at io.vertx.sqlclient.impl.command.CommandResponse.fire(CommandResponse.java:46)
        at io.vertx.sqlclient.impl.SocketConnectionBase.handleMessage(SocketConnectionBase.java:324)
        at io.vertx.pgclient.impl.PgSocketConnection.handleMessage(PgSocketConnection.java:114)
        at io.vertx.sqlclient.impl.SocketConnectionBase.lambda$init$0(SocketConnectionBase.java:137)
        at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:342)
        at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:335)
        at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:389)
        at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:159)
        at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
        at io.vertx.pgclient.impl.codec.PgDecoder.fireCommandResponse(PgDecoder.java:52)
        at io.vertx.pgclient.impl.codec.PgCommandCodec.handleReadyForQuery(PgCommandCodec.java:133)
        at io.vertx.pgclient.impl.codec.PgDecoder.decodeReadyForQuery(PgDecoder.java:248)
        at io.vertx.pgclient.impl.codec.PgDecoder.channelRead(PgDecoder.java:107)
        at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:796)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:732)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:658)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1583)

````

None of the workarounds mentioned above seem to work in my use case :(
